### ISSUE TITLE
SSH server configuration setup documentation

### ISSUE THREAD

**Issue by kgmyatthu** [directvt/vtm#401]
Sorry if this has been mentioned else where, it'd be great if someone could made a documentation on setting up vtm on a server and connect through local machine. It'd be awesome if people could use VTM to work on their remote machine. It's like SSH on steroid .

**Comment by o-sdn-o** [comment]
I fully support this idea, I need to describe it in the form of use cases in readme.md or in a standalone file at vtm//src/doc/.

**Comment by Vosjedev** [comment]
I guess I can write this while waiting for panel integration.

**Comment by o-sdn-o** [comment]
Pull requests are welcome!

**Comment by Vosjedev** [comment]
I'l work on it tomorrow.

**Comment by Vosjedev** [comment]
I guess this is the docs?

---

configure on the server:
put the following code in your .profile:
```bash
# (server .profile)
# launch vtm when using ssh
if [[ -v SSH_CONNECTION ]]; then vtm; fi
```
You wont be able to use 
```sh
ssh -t user@example.com bash
```
to get a bash shell instead, as bash sources your .profile. You could modify the code on the server to
```bash
# (server .profile)
# launch vtm when using ssh
if [[ -v SSH_CONNECTION ]] && ! [[ -v SSH_NO_VTM ]] ; then vtm; fi
```
and then use
```sh
ssh -t user@example.com SSH_NO_VTM='' bash
```
as command to connect to the server to overwrite the line.

---

configure on client side:
use 
```sh
ssh -t user@example.com vtm
```
instead of your default command, or
```sh
ssh -t user@example.com 'bash -c "source ~/.bashrc && vtm"'
```
to still have the environment variables from your .bashrc exported. 

---

if this is right, I will make a pr.

**Comment by Vosjedev** [comment]
I get this output while testing:
```
@ ssh -t vosje@192.168.178.125
(vosje@192.168.178.125) Password: 
(vosje@192.168.178.125) Verification code: 
Linux raspberrypi 6.1.19-v8+ #1637 SMP PREEMPT Tue Mar 14 11:11:47 GMT 2023 aarch64

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
Web console: https://raspberrypi:9090/ or https://192.168.178.125:9090/

You have new mail.
Last login: Sun Jun 18 09:43:14 2023 from 192.168.178.95
  os: terminal type "xterm-256color"
  os: color mode: true-color
  os: mouse mode: VT-style
vtm v0.9.8e
apps: loading configuration from $VTM_CONFIG...
apps: loading configuration from ~/.config/vtm/settings.xml...
        reading configuration: '/home/vosje/.config/vtm/settings.xml'
╭ ~ vosje@raspberrypi 0 -<
╰<%
```

**Comment by o-sdn-o** [comment]
Try public key authentication when logging in via SSH and make vtm the default shell with usermod --shell. In this case, you can connect via ssh user@server. I will describe this in more detail a little later.

**Comment by Vosjedev** [comment]
ah, forgot that (I am actualy using that already on my second server, but for a local tty instead of ssh, and using a wrapper).
But this makes vtm the default shell for all sessions, not only ssh right?

**Comment by o-sdn-o** [comment]
Yes it is.

**Comment by Vosjedev** [comment]
I use a wrapper on my server to do this, and the wrapper asks what shell to use on login. if it can find vtm in the PATH, it is preferred as default. It is a bash script, so it only works on linux or in things like cygwin or msys2. I have set my wrapper as default by modifying /etc/passwd with nano.

for a doc, I guess I will look into `usermod --shell.`
With public key auth, thats more a quality of life feature right? Or can you set a shell with it?

I also just thougt about adding vtm to /etc/shells, to be able to use chsh for it.

**Comment by o-sdn-o** [comment]
Those sessions that are launched via the vtm menu must be configured to use a normal shell.

**Comment by Vosjedev** [comment]
Yes, thats what I wrote the wrapper for in the first place. And then it became more advanced.

**Comment by o-sdn-o** [comment]
> With public key auth, thats more a quality of life feature right? Or can you set a shell with it?

Perhaps there is an option to specify the shell when logging in. It needs to be dug up.

**Comment by Vosjedev** [comment]
yea, im reading an article on public keys right now.

**Comment by o-sdn-o** [comment]
How to connect using DirectVT https://github.com/netxs-group/vtm/issues/321#issuecomment-1399331280

**Comment by o-sdn-o** [comment]
When using the direcvt protocol, the output occurs with the highest possible performance, since the binary stream is transmitted directly to the terminal and VT sequences are not used.

**Comment by o-sdn-o** [comment]
It seems to me that it is necessary to describe two modes of remote operation of vtm, without affecting the change in the default system and shell settings.

Two modes of operation of vtm via SSH:
- Normal VT mode, where I/O is done using VT sequences. The data is wrapped into VT-sequences on one side and parsed on the other. This is a relatively resource intensive process.
- DirectVT mode, in which data exchange takes place in binary form directly in both input and output without intermediate encoding and parsing of VT sequences.

Normal VT mode is available from any terminal and any shell, but it's slow. Launching vtm can be done in the same way as with any regular console program.

DirectVT (dtvt) is only available from the vtm environment and only when launched through a pre-configured menu item. In order to run vtm in DirectVT mode, you must connect to a remote server, excluding the use of interactive authentication, since interactive authentication always occurs in text mode, which is incompatible with dtvt. This can only be done using public-key authentication (or not using authentication at all, for example, for local cases).

**Comment by o-sdn-o** [comment]
Remote host:
- Make symlink to vtm in /usr/local/bin.

Local host:
- `./authkey`:
  ```
  your key to auth on remote host
  ```
- `~/.config/vtm/settings.xml`:
  ```xml
  <config>
    <menu>
      <item id="vtm over ssh" type=DirectVT param="ssh -q -i ./authkey user@server vtm"/>
    </menu>
  </config>
  ```

Start vtm on the local host and connect to the remote host via the "vtm over ssh" menu item.

**Comment by o-sdn-o** [comment]
vtm lacks command line option to launch programs in dtvt mode.

Something like
```
vtm --directvt "ssh -q -i ./authkey user@server vtm"
```

**Comment by o-sdn-o** [comment]
I'm sorry for the chaotic comments earlier today, because I didn't fully dive into the topic.

**Comment by o-sdn-o** [comment]
I have added "DirectVT Proxy Console". See `dev\sdn` branch.

It can be activated via "-r" option. All symbols following after `vtm -r directvt` are treated as command line with arguments.

For example:
- Run local app with DirectVT
  ```
  vtm -r directvt vtm -r truecolor
  ```
- Connect to remote host via SSH+DirectVT
  ```
  vtm -r directvt ssh -q -i ./authkey user@server vtm
  ```

**Comment by Vosjedev** [comment]
Thats very useful! Is it done? Should I make a doc somewhere? Maybe utilize githubs wiki feature?

**Comment by Vosjedev** [comment]
I will do a dev branch install on my computer to test with it.

**Comment by Vosjedev** [comment]
In the closed issue you linked earlier I found this:

> Note. SSH connection requires key-based authentication (w/o password protection) to avoid interactivity.

Maybe we have to add an ssh client to the app roadmap

**Comment by o-sdn-o** [comment]
> Is it done?

This still needs to be tested, there will probably be inconsistencies and they will need to be fixed. Also, because of my incorrect branching, these changes are in the same branch as ngvt-input-mode, so this will be merged into the main branch in bulk. I'll fix my branching after next merge.

> Should I make a doc somewhere?

Yes, it would be very good to describe it. You can describe this particular feature as a description of the built-in DirectVT Proxy Console application in the `vtm/doc/apps.md` file, or, if you will, you can describe in general how vtm works both locally and remotely:
- Describe a server instance that allows multi-user remote and local connections that exists as a daemon on the system.
- Describe that, by default, a server instance is tied to the user account that launched it.
- Describe that you can run several of these daemons using different pipe names `vtm -p another_pipe_name`, and they will exist in parallel and accept connections.
- Any such daemon can be monitored with `vtm -m`.
- Describe how to connect to an already running server instance, both locally and remotely via SSH. If the server instance is not running, then it automatically starts for this user.
- Describe the advantages of the DirectVT protocol in that it eliminates the need for an intermediate transcoding of VT sequences.

The main emphasis should be on avoiding or minimizing customization of the user system. We'd better reimplement vtm's behavior if such changes are needed.

It can be described that vtm has the same goals as tmux, just in a slightly different way.

P.S.: My problem with this is that my English does not allow me to express myself clearly yet, I can only express myself in a list of theses, avoiding long sentences. I can't distinguish a situation where people don't understand what I wrote, from a situation where people understand everything but disagree. In both cases there is no feedback and this slows down my progress a bit.

> Maybe utilize githubs wiki feature?

Would be better to make it as a separate file, something like `vtm/doc/guide.md` (concept.md, about.md, usage.md). We will later change the name of the document according to its contents.

**Comment by o-sdn-o** [comment]
> Maybe we have to add an ssh client to the app roadmap

I decided to load the vtm codebase with external dependencies to a minimum. At the moment, all operating systems have a regular ssh-client and we can use it.

**Comment by o-sdn-o** [comment]
I mean this section of the file `vt/doc/apps.md`

https://github.com/netxs-group/vtm/blob/b841b67de28f311abe7094bbe4e890ff47242b5c/doc/apps.md?plain=1#L1-L6

https://github.com/netxs-group/vtm/blob/b841b67de28f311abe7094bbe4e890ff47242b5c/doc/apps.md?plain=1#L232-L236

**Comment by Vosjedev** [comment]
> This still needs to be tested


I can help with that.

> My problem with this is that my English does not allow me to express myself clearly yet

no problem, I would love to test my english by writing this. For me this is both helping you and my own english.

> Would be better to make it as a separate file

oke, and then link to it from the readme? Or should I make a file linking to all docs and link that in the readme?

**Comment by Vosjedev** [comment]
> I decided to load the vtm codebase with external dependencies to a minimum. At the moment, all operating systems have a regular ssh-client and we can use it.

yea, that's understandable

**Comment by o-sdn-o** [comment]
The keyword "directvt" can be replaced with the short "dtvt".

**Comment by o-sdn-o** [comment]
https://github.com/netxs-group/vtm/assets/11535558/10fe955c-9041-41aa-ad0d-a6a6698d8ed5

**Comment by Vosjedev** [comment]
I saw that in the realease notes, I will include it in the docs. VTM can talk via any tpc socket right? Can it talk via a named pipe (fifo)?

**Comment by o-sdn-o** [comment]
For vtm any redirected channel is suitable. It just reads stdin and sends to stdout.

**Comment by o-sdn-o** [comment]
> Can it talk via 2 named pipes (fifo) (one for server to client, one for client to server? I guess this is only 1 client talking via the pipe.

I can't understand this mechanic. Can you give an example?

**Comment by Vosjedev** [comment]
I tried this:
```sh
mkfifo in && mkfifo out
tail -f in | vtm >> out # start the server
```
And then to connect:
```sh
tail -f out | vtm >> in
```
But it does not seam to do anything.

**Comment by Vosjedev** [comment]
Also when trying the ncat example:

https://github.com/netxs-group/vtm/assets/100359465/4676c05d-a22a-4ecc-b742-9a1fd6965f2d

I have some issues:
- I can only move the viewport around by one pixel at a time until I release mouse
- (not shown in video) vtm not responding anymore on disconnect, and then showing 'no signal' on screen.

**Comment by o-sdn-o** [comment]
> I am redirecting vtm's output to another vtm's input, and the output of that one to the input of the first one.

I forgot to clarify that redirected I/O only works inside DirectVT, that is, only when vtm is running inside `vtm -r dtvt`.

```bash
vtm -r dtvt _run_command_in_dtvt_mode_
```

**Comment by o-sdn-o** [comment]
> I can only move the viewport around by one pixel at a time until I release mouse.

Tomorrow I will try to reproduce this behavior and fix it.

> vtm not responding anymore on disconnect, and then showing 'no signal' on screen. When I tried to record this, it would not happen.

I also encountered this behavior. I tried today to determine the source of this behavior, but so far without success. This is like a file descriptor being leaked during a fork. This is reproduced when using nc instead of ncat on the client side.

**Comment by o-sdn-o** [comment]
> I can only move the viewport around by one pixel at a time until I release mouse.

Fastest repro:
```
vtm -r dtvt vtm
```

**Comment by o-sdn-o** [comment]
> I tried this:
> ```bash
> mkfifo in && mkfifo out
> tail -f in | vtm >> out # start the server
> ```

In the case of using two unidirectional channels, we need to redirect them asynchronously to standard input/output. `Socat` is good for that:

Server side:
```bash
mkfifo in && mkfifo out
vtm >out <in
```

Client side:
```bash
vtm -r dtvt socat open:out\!\!open:in stdio
```

**Comment by Vosjedev** [comment]
ahhh, that explains my example not working.

I guess this is also a good example to include in the docs about dtvt

**Comment by o-sdn-o** [comment]
I have not yet investigated the source of the environment shutdown hang when using netcat. Perhaps there will be some workaround.

**Comment by Vosjedev** [comment]
yea, I have some issues. The ubuntu updater bricked python to the point where xonsh is unusable, so I am trying to get that back first.

**Comment by o-sdn-o** [comment]
Another useful option is to use `inetd` to share vtm or dtvt-apps on LAN

- Server side
  - Install inetd
  - Update /etc/inetd.conf:
    ```
    1234 stream tcp nowait root /usr/local/bin/vtm vtm
    ```
  - Launch `inetd`
- Client side
  ```
  vtm -r dtvt ncat server_ip 1234
  ```

**Comment by o-sdn-o** [comment]
It is important to emphasize that some implementations of netcat (e.g. `nc` on Ubuntu 22.04) do not terminate on the client side even though the process terminates on the server side.

Repro:
- Server
  ```bash
  ncat -l 1234 -ke '/usr/bin/cat'
  ```
- Client
  ```bash
  nc 127.0.0.1 1234
  ```
- Press Ctrl+C on Server
- Client is waiting forever

**Comment by o-sdn-o** [comment]
Here is what we have (maybe there is something else)

# Standard I/O Redirection

- Server
  ```bash
  mkfifo in && mkfifo out
  vtm >out <in
  ```
- Client:
  ```bash
  vtm -r dtvt socat open:out\!\!open:in stdin\!\!stdout
  # `vtm -r dtvt` to run DirectVT proxy.
  # Note: Make sure `socat` is installed.
  ```

# Multiuser Remote Access Scenarios

## Using SSH
  - Server:
    - Install SSH-server
  - Client:
    ```bash
    ssh -t user@server /server/side/path/to/vtm
    # `-t` to allocate TTY.
    ```

## Using SSH + DirectVT
  - Server:
    - Install SSH-server
    - Configure key-based authentication for SSH
  - Client:
    ```bash
    vtm -r dtvt ssh -q -i /path/to/auth_keys user@server /server/side/path/to/vtm
    # `vtm -r dtvt` to run DirectVT proxy.
    # `-q` to suppress ssh texting in binary stream.
    # `-i` to specify auth keys.
    ```
## Using `netcat` + DirectVT
  - Server:
    ```bash
    ncat -l server_port -k -e /server/side/path/to/vtm
    # `-l server_port` to specify tcp port to listen.
    # `-k` to keep open for multiple clients.
    # `-e` to run vtm for every connected client.
    # Note: Make sure `ncat` is installed.
    ```
  - Client:
    ```bash
    vtm -r dtvt ncat server_ip server_port
    # `vtm -r dtvt` to run DirectVT proxy.
    # Note: Make sure `ncat` is installed.
    ```
## Using `inetd` + DirectVT
  - Server:
    - Install `inetd`
    - Add the following line to the `/etc/inetd.conf`:
      ```bash
      server_port stream tcp nowait user_name /server/side/path/to/vtm  vtm
      # `server_port` to specify tcp port to listen.
      # `user_name` to specify user login name.
      ```
    - Launch `inetd`
      ```
      inetd
      ```
  - Client
    ```bash
    vtm -r dtvt ncat server_ip server_port
    # `vtm -r dtvt` to run DirectVT proxy.
    # Note: Make sure `ncat` is installed.
    ```

Note: Some netcat implementations (e.g. nc (netcat-openbsd) on Ubuntu 22.04) don't terminate on the client side even though the process terminates on the server side.

**Comment by Vosjedev** [comment]
wow, I guess you already made a doc before me. I am waitin way to long.

You asked me to describe the dtvt protocol, should I add that to this?

**Comment by o-sdn-o** [comment]
This is a sketch to have something to rely on. There is no description of setting up connections through the menu in the vtm environment. There is no statement about the practical meaning of such connections. I mean use cases like in tmux for example. We need to think about another week. We will collect ideas and make a concise description of them. Perhaps vtm will need to be corrected yet. You can just write ideas here for now, and we will combine them all. We have a lot of time.

**Comment by o-sdn-o** [comment]
Such scenarios have not been practiced by anyone for TUI applications, and use cases are possible, both opening up new opportunities and opening security issues. Sort of a [reverse shell](https://wiki.ubuntu.com/ReverseShell).

**Comment by o-sdn-o** [comment]
On the other hand, it is possible to publish applications that do not allow the execution of commands on the server, but allow users to interact with each other on the server site.

**Comment by o-sdn-o** [comment]
> You asked me to describe the dtvt protocol, should I add that to this?

At the beginning, it will be necessary to place a description of the DirectVT protocol. Without this, the description of use cases will not be clear.

There are three main goals of dtvt:
- Operation of a TUI application without an allocated TTY.
- Getting rid of transcoding VT-sequences back and forth.
- Deep integration of the application into the vtm environment, giving insane smoothness when resizing and support for multi-user UI.
Multi-user UI is when two users can resize a window at the same time by dragging the edges.

**Comment by o-sdn-o** [comment]
I have added a description of using via SSH in [/doc/architecture.md](https://github.com/directvt/vtm/blob/master/doc/architecture.md)

**Comment by o-sdn-o** [comment]
I think the description that we ended up with in the form of the /doc/architecture.md file is quite sufficient to describe remote access. If there is something missing and more needs to be added, feel free to re-open this issue.
